include/picongpu/particles/creation/ModuleConfig.hpp

```diff
@@ -0,0 +1,102 @@
+     *   - the atomicNumbers of the chargeStateDataDataBox and atomicStateDataDataBox passed via additionalData are
+     *     consistent
+     * @tparam T_SkipSuperCellFunctor test allowing entire superCell to be skipped depending on additionalData or
+     * superCell index
+     * e.g. skip superCell if localTimeRemainingDataBox[sharedDataBoxIndex] is > 0
+     * @tparam T_PredictorFunctor functor predicting number of product species particles to spawn for a given source
+     * species particle, depending on sharedState and additionalData.
+     * @note may update source particle!
+     * @tparam T_ParticlePairUpdateFunctor functor initialising spawned productSpecies particle based on
+     * additionalData and sourceSpecies particle
+     * @tparam T_SharedStateType type of sharedState, one instance for each superCell
+     * @tparam T_InitSharedStateFunctor functor initialising sharedState variable
+     * @tparam T_AdditionalDataIndexFunctor functor returning index to access additionalData by
+     *  @note only one is supported for all additionalData
+     *  @note dimension is configurable
+     *  @note may be ignored for some or all additionalData
+     * @tparam T_WriteOutSharedStateFunctor write out final sharedState
+     */
+    template<
+        template<
+            typename T_SourceParticleBox,
+            typename T_ProductParticleBox,
+            typename... T_KernelConfigOptionsAndSharedDataBoxTypes>
+        typename T_SanityCheckInputs,
+        template<typename... T_KernelConfigOptions>
+        typename T_SkipSuperCellFunctor,
+        template<typename T_Number, typename... T_KernelConfigOptions>
+        typename T_PredictorFunctor,
+        typename T_ParticlePairUpdateFunctor,
+        typename T_SharedStateType,
+        template<typename... T_KernelConfigOptions>
+        typename T_InitSharedStateFunctor,
+        template<typename... T_KernelConfigOptions>
+        typename T_SharedDataBoxIndexFunctor,
+        template<typename... T_KernelConfigOptions>
+        typename T_WriteOutSharedStateFunctor>
+    struct ModuleConfig
+    {
+        template<
+            typename T_SourceParticleBox,
+            typename T_ProductParticleBox,
+            typename... T_KernelConfigOptionsAndSharedDataBoxTypes>
+        using SanityCheckInputs = T_SanityCheckInputs<
+            T_SourceParticleBox,
+            T_ProductParticleBox,
+            T_KernelConfigOptionsAndSharedDataBoxTypes...>;
+
+        template<typename... T_KernelConfigOptions>
+        using SkipSuperCellFunctor = T_SkipSuperCellFunctor<T_KernelConfigOptions...>;
+
+        template<typename T_Number, typename... T_KernelConfigOptions>
+        using PredictorFunctor = T_PredictorFunctor<T_Number, T_KernelConfigOptions...>;
+
+        template<typename... T_KernelConfigOptions>
+        using ParticlePairUpdateFunctor = T_ParticlePairUpdateFunctor<T_KernelConfigOptions...>;
+
+        using SharedStateType = T_SharedStateType;
```

Comment:
`SharedKernelContext`

Reply:
renamed to KernelStateType as discussed offline